src/Generic/Data/Internal/Error.hs

```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Generic.Data.Internal.Error where
+
+import GHC.Generics
+import GHC.TypeLits
+
+type family Or a b where
+  Or 'False 'False = 'False
+  Or 'False 'True = 'True
+  Or 'True 'False = 'True
+  Or 'True 'True = 'True
+
+type family Not a where
+  Not 'True = 'False
+  Not 'False = 'True
+
+type family HasSum f where
+  HasSum V1 = 'False
+  HasSum U1 = 'False
+  HasSum (K1 i c) = 'False
+  HasSum (M1 i c f) = HasSum f
+  HasSum (f :*: g) = Or (HasSum f) (HasSum g)
+  HasSum (f :+: g) = 'True
+
+type family AssertH pred msg where
+  AssertH 'True msg = '()
+  AssertH 'False msg = TypeError msg
```

Comment:
This family could produce a  `Constraint` instead of a `Type`, so that you can write only `AssertH pred msg` instead of `AssertH pred msg ~ '()`. In fact it would take away the need for a separate `Assert` class.

Reply:
That's what I tried first, but unfortunately a simple constraint is deferred to the usage of the instance, and not its declaration, resulting with the previous error.

I'm not sure why but the type equality at instance declaration trick seems to be necessary. I tried to find clues in the documentation of type equality and type families but didn't come up with an explanation..

Reply:
How curious. I don't know what's going on either. In any case, this can be simplified to only the class and two instances:

```
class Assert (pred :: Bool) (msg :: ErrorMessage)
instance Assert 'True msg
instance (TypeError msg ~ '()) => Assert 'False msg
```

(It's also possible with only the family by producing the `TypeError msg ~ '()` constraint directly, but that feels slightly more complicated.)

Reply:
Good suggestion, done!

Regarding what is going on, I figured that the common use case for TypeError is to translate some "bad instance" to a type error, so it deliberately doesn't trigger in instances, but that by using it as a type family parameter or in a type equality we "force" it to trigger at this stage.

Reply:
Oh, that makes sense!